runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -993,29 +1010,29 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 	 * Checks that the dataAddr field of the indexable object is correct.
 	 *
 	 * @param arrayPtr      Pointer to the indexable object
+	 * @param isValidDataAddrForDoubleMappedObject	Boolean to determine whether the given indexable object is double mapped
 	 * @return if the dataAddr field of the indexable object is correct
 	 */
 	MMINLINE bool
-	isValidDataAddr(J9IndexableObject *arrayPtr)
+	isValidDataAddr(J9IndexableObject *arrayPtr, bool isValidDataAddrForDoubleMappedObject)
 	{
 		bool isValidDataAddress = true;
-
 		if (_isIndexableDataAddrPresent) {
 			void *dataAddr = getDataAddrForIndexableObject(arrayPtr);
-			isValidDataAddress = isValidDataAddr(arrayPtr, dataAddr);
+			isValidDataAddress = isValidDataAddr(arrayPtr, dataAddr, isValidDataAddrForDoubleMappedObject);
 		}
-
 		return isValidDataAddress;
 	}
 
 	/**
 	 * Checks that the dataAddr field of the indexable object is correct.
 	 *
 	 * @param arrayPtr      Pointer to the indexable object
+	 * @param isValidDataAddrForDoubleMappedObject	Boolean to determine whether the given indexable object is double mapped
 	 * @return if the dataAddr field of the indexable object is correct
 	 */
 	MMINLINE bool
-	isValidDataAddr(J9IndexableObject *arrayPtr, void *dataAddr)
+	isValidDataAddr(J9IndexableObject *arrayPtr, void *dataAddr, bool isValidDataAddrForDoubleMappedObject)
 	{
 		bool isValidDataAddress = false;
 		uintptr_t dataSizeInBytes = getDataSizeInBytes(arrayPtr);
```

Comment:
if dataSizeInBytes 0 case, the check could be more specific
- if 0 sized array, dataAddr is null
- if non-0 size, dataAddr is end of header

the point not to use 'or', because if it's non-0 size and dataAddr is null, the check would pass, but it should not

Reply:
the above comment might be stale now, but what we should do is to simplify the check and not allow dataAddr null (it's only possible when offheap is disabled, but this check should only be called when offheap is enabled)